### ISSUE TITLE
.luacheckrc in parent hierarchy

### ISSUE THREAD

**Issue by za-creature** [mpeterv/luacheck#33]
A lot of linters (jshint, jscs, scss-lint, coffeelint) that allow configuration via a .rc file look for it in the folder of the file being linted, but will also look in all parent folders until either a .rc file is found or the root of the filesystem is reached.

I see no reason why luacheck couldn't do the same thing. The performance impact is minimal and has the added benefit of being familiar with the above mentioned linters, which is nice for web projects that use lua either as a backend, or to leverage Redis' scripting functionality.

In my opinion, the decision of looking for `.luacheckrc` in the current working directory is questionable at best; the whole point of the config file is to be coupled with the source tree (for example to add per-project globals), whereas if using absolute paths, you'll get different behavior depending on where exactly you're running `luacheck` from. I fail to see a use case when this is what you want.

**Comment by mpeterv** [comment]
I agree that implementing this would be an improvement, but I don't think it's that big. Advantages are that you don't have to run luacheck from within project directory and that you can check several projects at once. Also you could use several configs for different subdirectories within a project, but per-path overrides handle that already.

So while luacheck's behaviour is not familiar, it works similarly to linters you mentioned in case you run it from any directory within the project. Since some other tools, like git, have the same restriction, I think the problem is not significant. @ignacio @xpol what do you think?

**Comment by xpol** [comment]
I think the rc file look up should stop at project root directory.
And the config file in user home directory should take in count.

luacheck may support more than one command line specified rc files.

**Comment by za-creature** [comment]
@mpeterv Not really similar. [SublimeLinter-lua](https://github.com/SublimeLinter/SublimeLinter-lua) runs it relative to the input file's directory and as such cannot be used unless there's a `.lualintrc` (or link to) in every folder containing .lua files.

@xpol How exactly do you define 'project root directory'?

**Comment by mpeterv** [comment]
@za-creature I'm sorry, I think I misunderstood you. luacheck 0.10.0 only looks for config in the current directory, but on master branch for quite some time it also recursively checks all upper directories, see #20. It's not the same as starting from the checked file's directory, but it works with ST3 plugin, and as I said it works if you start luacheck from any directory within the project.

**Comment by za-creature** [comment]
@mpeterv

Apologies then; you may close this if you wish. Any estimate as to when the next point release is due?

I still disagree with the semantics; from my perspective, your project violates the [POTA](https://en.wikipedia.org/wiki/Principle_of_least_astonishment) as the userbase is divided into two categories:

1. People who haven't used a linter before and who would be fine with any semantics as they're not expecting anything.
2. People who have used a linter before and are expecting compatibility with the existing status quo (to what I could find on github, luacheck is about a year old, which is way younger than the most popular linters). Of course, there's nothing stopping you from going against the status quo, but my guess is that since you've open sourced the project, you're aiming for maximum usefulness.

The analogy to git is also rather poor, since after you take into account the fact that both git and luacheck operate on files and both provide a command line user interface, the similarities end. Git has a notion of a repository (a folder containing a .git subfolder) which in most cases is analogous to a project (or subproject), whereas luacheck by definition must be agnostic to anything that isn't lua. 

The general UNIX consensus is that all command line tools completely ignore the working directory except for path resolution purposes (going as far as removing `.` from `PATH`, thereby forcing you to explicitly type `./` to confirm you're *absolutely sure* you want to run a local executable). Git violates this because it's project-centric and completely disregards anything that's outside of its root folder; as such, it makes sense for its user interface to be relative to the current working directory. Unfortunately, luacheck does not have that privilege, as it does not have the notion of a project (checking for the presence of a `.git` subfolder is a hack at best).

In my opinion, a better analogy to `luacheck` would be either `cat` or `stat`: all 3 commands should accept one or more filenames, provide some (existing or computed) data related to the input files (and any extra options passed on), and said data should be the same, regardless of the current working directory.

A simple example of where this breaks (even if you're going for a project-centric git-like user interface, which once again, you cannot as you can't assume your users are using git and not mercurial for example):

```
project/
    app.lua
    redis/
        .luacheckrc
        .script1.lua
        .script2.lua
    backend/
        .luacheckrc
        handler1.lua
        handler2.lua
    .luacheckrc
```
`luacheck redis/script1.lua` from `project/` will use `project/.luacheckrc`
`luacheck script1.lua` from `project/redis` will use `project/redis/.luacheckrc`
`luacheck ../redis/script1.lua` from `project/backend` will use `project/backend/.luacheckrc`

Disclaimer: I have 0 knowledge of using Lua outside of Redis, but the example should still be relevant.

**Comment by mpeterv** [comment]
@za-creature the point about git is: because some tools, like git and luarocks, force you to `cd` into your project first, you are most likely to only ever run luacheck from within the project, and then there is no difference between looking for config starting from current directory or file directory, unless for some reason you want to check another project.

The idea so far is that if you want to have different options for different directories, use per-path overrides in the root config, so there is only one config per project. jshint also supports this with "overrides" option.

I'm not arguing that your suggestion isn't an improvement, but that approach currently implemented on master branch is still quite flexible, and it works with linters, so as long as it's documented, this won't be a critical issue. I'd merge a PR implementing the better method, but won't work on it myself yet.

0.11.0 will be released next week (13.07 - 17.07).

**Comment by za-creature** [comment]
Unfortunately, I'm not confident enough in my general-purpose Lua knowledge to contribute (I just use it for redis scripting) so that I'll leave that PR to someone else.

**Comment by mpeterv** [comment]
0.11.0 released.

**Comment by jasonwilliams200OK** [comment]
+1
The whole story goes like this:

* Look for rc file in the current directory.
* Then parent directory.
* Then grand parent and so on.. until you hit the drive root.
* Penultimately, check in HOME directory (unix: `~/` or win: `%userprofile%`)
* Finally, load defaults with "config not found" yellow warning (or throw red error and crash)
